### PR TITLE
feat: endpoints to link spotify account and get your currently playing song

### PR DESCRIPTION
**Pull request by sriramb2000**
Note: the way the linking is done is kinda jank - rn the link route returns a url, that you can then visit to authorize the spotify linking. There's cors bullshit you have to deal with if you do a redirect from the server side and I don't want to think about that so we're going with this jank method.

### CODE REVIEW

File: backend/src/routes/index.js [hunk 1]
```diff
@@ -2,11 +2,13 @@ const express = require('express');
 
 const { authMiddleware, getAuthRoutes } = require('./auth.js');
 const { getSongRoutes } = require('./songs.js');
+const { getSpotifyRoutes } = require('./spotify');
```
Comment (sjiang647) [inline backend/src/routes/index.js:5]: nit: make these 3 imports the same b/c ocd

**Review by sjiang647 — APPROVED** [review]
Some nits

File: backend/src/routes/index.js [hunk 2]
```diff
@@ -2,11 +2,13 @@ const express = require('express');
...
 
 function getRoutes() {
   const router = express.Router();
   router.use('/auth', getAuthRoutes());
   router.use('/songs', authMiddleware, getSongRoutes());
+  router.use('/spotify', getSpotifyRoutes());
```
Comment (michaelw54) [inline backend/src/routes/index.js:11]: Could you add the auth middleware to the spotify route? We wouldn't want people linking spotify accounts to nothing...

File: backend/src/routes/spotify.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+const { firebase } = require('../firebase-init.js');
+const { authMiddleware } = require('./auth');
+
+const db = firebase.firestore();
+require('dotenv').config();
+
+const Spotify = new SpotifyWebApi({
+  clientId: process.env.SPOTIFY_CLIENT_ID,
+  clientSecret: process.env.SPOTIFY_CLIENT_SECRET,
+  redirectUri: 'https://hopscotch-backend.herokuapp.com/api/spotify/oauth-callback',
```
Comment (michaelw54) [inline backend/src/routes/spotify.js:13]: Should we put all instances of internal URLs into an env file? So when we deploy we use an env file w say
```
BASE_URL=https://hopscotch-backend.herokuapp.com
```
and for testing we use
```
BASE_URL=http://localhost:5000
```
and then
```
redirectUri: `${process.env.BASE_URL}/api/spotify/oauth-callback`
```
non-blocking, but let's do this on a separate PR

Edit: realized you put redirectHost as an optional param in `handleLink`. This could work, but let's not have the user decide where to redirect to.

File: backend/src/routes/spotify.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+Creates a url that can be visited by the client to link Spotify account with
+your Hopscotch account.
+*/
+async function handleLink(req, res) {
+  const state = crypto.randomBytes(20).toString('hex');
+  // Store the state in a document for lookup later, store the corresponding user's email
+  await db.collection('spotify_state').doc(state.toString()).set({
+    email: req.user_email,
+    createdAt: firebase.firestore.FieldValue.serverTimestamp(),
+  });
```
[the remark this answers is not in the document]
Comment (sriramb2000) [reply backend/src/routes/spotify.js:47]: `.set()` creates the document if it doesn't exist or overwrites it if it does exist so `.set()` works here.

File: backend/src/routes/spotify.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+  if (req.query.redirectHost) {
+    Spotify.setRedirectURI(`${req.query.redirectHost}/api/spotify/oauth-callback`);
+  }
+  const authorizeURL = Spotify.createAuthorizeURL(SPOTIFY_SCOPES, state.toString());
+  res.json({ authorizeURL });
```
[the remark this answers is not in the document]
Comment (sriramb2000) [reply backend/src/routes/spotify.js:53]: linter didn't actually give me any error for this but i'll go ahead and tweak it

File: backend/src/routes/spotify.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  'playlist-read-collaborative',
+  'user-read-playback-state',
+  'user-read-currently-playing',
+  'user-library-read',
+  'user-top-read',
+];
+
+/*
+-- Request Params --
+- redirectHost (OPTIONAL) -- If you are attempting to run this locally, you can pass in your local server host (e.g. "http://localhost:5000")
```
Comment (michaelw54) [inline backend/src/routes/spotify.js:31]: Should we leave the redirectHost up to the user? Related to the comment I left above

**Review by michaelw54 — APPROVED** [review]
LGTM with minor comments

File: backend/src/routes/spotify.js (hunk 6 above)
Comment (sriramb2000) [reply backend/src/routes/spotify.js:31]: redirectHost only works if the redirect url actually matches smth that I've manually registered in the Spotify Developer Console, but yea we can just go w the base url thing
